File: daily_python/2337_Move_Pieces_to_Obtain_a_String.py
import os
import sys
from typing import List, Optional

class Solution:
    def canChange(self, start: str, target: str) -> bool:
        n = len(start)
        start_idx = target_idx = 0
        while start_idx < n or target_idx < n:
            while start_idx < n and start[start_idx] == '_':
                start_idx += 1
            while target_idx < n and target[target_idx] == '_':
                target_idx += 1
            if start_idx == n or target_idx == n:
                return start_idx == target_idx
            if start[start_idx] != target[target_idx] or (start[start_idx] == 'L' and target_idx > start_idx) or (start[start_idx] == 'R' and start_idx > target_idx):
                return False
            start_idx += 1
            target_idx += 1
        return True
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            start = lines[0].split('=')[1].strip()[1:-2]
            target = lines[1].split('=')[1].strip()[1:-2]
            dataset.append((start, target))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for start, target in dataset:
        results.append(solution.canChange(start, target))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2024/2337_Move_Pieces_to_Obtain_a_String.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    bool canChange(string start, string target) {
        int n = start.length();
        int start_idx = 0;
        int target_idx = 0;
        while(start_idx < n || target_idx < n){
            while(start_idx < n && start[start_idx] == '_') start_idx++;
            while(target_idx < n && target[target_idx] == '_') target_idx++;
            if (start_idx == n || target_idx == n) return start_idx == target_idx;
            if ((start[start_idx] != target[target_idx]) || (start[start_idx] == 'L' && target_idx > start_idx) || (start[start_idx] == 'R' && start_idx > target_idx)) return false;
            start_idx++, target_idx++;
        }
        return true;
    }
};

int main(){
    string start = "_L__R__R_";
    string end = "L______RR";
    // string start = "R_L_";
    // string end = "__LR";
    // string start = "_R";
    // string end = "R_";
    // string start = "_LL__R__R_";
    // string end = "L___L___RR";
    // string start = "R_L__R__R_";
    // string end = "_L______RR";
    // string start = "_L";
    // string end = "LR";
    // string start = "R";
    // string end = "_";
    Solution S;
    cout << S.canChange(start, end) << endl;
    return 0;
}